lua/lvim/updater.lua

```diff
@@ -0,0 +1,233 @@
+local M = {}
+
+local uv = vim.loop
+local packer = require "packer"
+local a = require "packer.async"
+local async = a.sync
+local await = a.wait
+
+local timer = {}
+function timer:start()
+  self.time = uv.hrtime()
+end
+function timer:stop()
+  return (uv.hrtime() - self.time) * 1e-6
+end
+timer.__index = timer
+
+local function get_lvim_after_user_config()
+  local original_lvim = lvim
+  local user_lvim = vim.deepcopy(lvim)
+  local original_package_loaded = package.loaded
+  local user_package_loaded = {}
+  _G.lvim = user_lvim
+  _G.package.loaded = user_package_loaded
+  local ok, err = pcall(dofile, require("lvim.config"):get_user_config_path())
+  if not ok then
+    print(err)
```

Comment:
I believe you misunderstand the use of this function, so it the variable swaps and the rest of it doesn't make sense to you. Think of it like a sandbox for the extensive damage the user config file does by default, it is inherently unpredictable and full of side-effects.

The user's `config.lua` file (as retrieved by get_user_config_path) can run arbitrary code. As such, it can and by default does mutate globals such as `_G.lvim` and `_G.package`. For the purpose of the updater it must, not influence the state of neovim, which [`config:load()` does extensively even after loading](https://github.com/LunarVim/LunarVim/blob/rolling/lua/lvim/config/init.lua#L82-L100). The user configured values for certain parts however, are valuable.

> currently re-assigning the values without even using it

This is incorrect, I think you may have misread the code or misunderstand the effect of what `dofile` is here. There are value swaps for globals that the user configuration mutates, but are integral to the updater script without allowing it to mutate everything else. For example, the state of core plugins, the user installed plugins, etc.

Should it use `getenv` and `setenv` for sandboxing behavior and sideffects? Absolutely, but the level of complexity to get that work work with how `lvim` is setup was far greater than this short function when I wrote this.

---

TLDR: it loads the user's configuration file in such a way that it does not mutate state, it is required and is not redundant. It's not missing error handling.